__test__/unit/services/indexedDb.test.ts

```diff
@@ -1,23 +1,105 @@
     await db.remove("Options", 'optionsKey');
     const retrievedValue = await db.get("Options", 'optionsKey');
     expect(retrievedValue).toBeUndefined();
   });
 });
+
+describe('migrations', () => {
+  describe('v5', () => {
+    test('can to write to new Outcomes.NotificationClicked table', async () => {
+      const db = newOSIndexedDb("testDbv5", 5);
+      const result = await db.put("Outcomes.NotificationClicked", { notificationId: "1" } );
+      expect(result).toEqual({ notificationId: "1" });
+    });
+
+    test('can write to new Outcomes.NotificationReceived table', async () => {
+      const db = newOSIndexedDb("testDbv5", 5);
+      const result = await db.put("Outcomes.NotificationReceived", { notificationId: "1" } );
+      expect(result).toEqual({ notificationId: "1" });
+    });
+
+    // Tests NotificationClicked records migrate over from a v15 SDK version
+    test('migrates notificationId type records into Outcomes.NotificationClicked', async () => {
+      const dbName = "testDbV4upgradeToV5" + Random.getRandomString(10);
+      const db = newOSIndexedDb(dbName, 4);
+      await db.put("NotificationClicked", { notificationId: "1" });
+      db.close();
+
+      const db2 = newOSIndexedDb(dbName, 5);
+      const result = await db2.getAll("Outcomes.NotificationClicked");
+      expect(result).toEqual([{appId: undefined, notificationId: "1", timestamp: undefined}]);
+    });
+
+    // Tests NotificationReceived records migrate over from a v15 SDK version
+    test('migrates notificationId type records into Outcomes.NotificationReceived', async () => {
+      const dbName = "testDbV4upgradeToV5" + Random.getRandomString(10);
+      const db = newOSIndexedDb(dbName, 4);
+      await db.put("NotificationReceived", { notificationId: "1" });
+      db.close();
+
+      const db2 = newOSIndexedDb(dbName, 5);
+      const result = await db2.getAll("Outcomes.NotificationReceived");
+      expect(result).toEqual([{appId: undefined, notificationId: "1", timestamp: undefined}]);
+    });
+
+    // Tests records coming from a broken SDK (160000.beta4 to 160000) and upgrading to fixed v5 db
+    test('migrates notification.id type records into Outcomes.NotificationClicked', async () => {
+      const dbName = "testDbV4upgradeToV5" + Random.getRandomString(10);
+
+      // 1. Put the db's schema into the broken v4 state that SDK v16000000 had
+      const openDbRequest = indexedDB.open(dbName, 4);
+      const dbOpenPromise = new Promise((resolve) => {
+        openDbRequest.onsuccess = resolve;
+      });
+      const dbUpgradePromise = new Promise<void>((resolve) => {
+        openDbRequest.onupgradeneeded = function (event: any) {
+          const target = event.target as IDBOpenDBRequest;
+          const db = target.result;
+          db.createObjectStore("NotificationClicked", { keyPath: "notification.id" });
+          db.createObjectStore("NotificationReceived", { keyPath: "notificationId" });

```

Comment:
what's the point of this line if we're not testing the migration for "NotificationReceived" below?

Reply:
Test is setting up the DB for how it should look for the v4 to v5 migration in production, even if we are not testing the NotificationReceived in this specific test. The test doesn't work without this line either, as the onupgradeneeded won't finish without as `migrateOutcomesNotificationReceivedTableForV5` throws.